spec/mail/messagesubmission.mdown

```diff
@@ -0,0 +1,311 @@
+# MessageSubmission
+
+The MessageSubmission object represents the submission of a message for delivery to one or more recipients. A **MessageSubmission** object has the following properties:
+
+- **id**: `String`
+  The id of the message submission. This is server-set and immutable.
+- **identityId**: `String`
+  The id of the identity to associate with this submission. This is immutable.
+- **messageId**: `String`
+  The id of the message to send. This is immutable. The message being sent does not have to be a draft, for example when "redirecting" an existing message to a different email address.
+- **threadId**: `String`
+  The thread id of the message to send. This is immutable and set by the server to the *threadId* property of the message referenced by the *messageId*.
+- **envelope**: `Envelope|null`
+  Information for use when sending via SMTP. This is immutable.
+
+    An **Envelope** object has the following properties:
+
+    - **mailFrom**: `Address`
+      The email address to use as the return address in the SMTP submission, plus any parameters to pass with the MAIL FROM address. The JMAP server MAY allow the email to be the empty string.
+
+        When a JMAP server performs a message submission, it MAY use the same id string for the [@!RFC3461] ENVID parameter and the MessageSubmission
+        object id. Servers that do this MAY replace a client-provided value  for ENVID with a server-provided value.
+
+    - **rcptTo**: `Address[]`
+      The email addresses to send the message to, and any RCPT TO parameters to pass with the recipient.
+
+    An **Address** object has the following properties:
+
+    - **email**: `String`
+      The email address being represented by the object. This as a "Mailbox" as used in the Reverse-path or Foward-path of the MAIL FROM or RCPT TO command in [@!RFC5321
+    - **parameters**: `Object|null`
+      Any parameters to send with the email (either mail-parameter or rcpt-parameter as appropriate, as specified in [@!RFC5321]). If supplied, each key in the object is a parameter name, and the value the parameter value. In both cases, any xtext or unitext encodings are removed ([@!RFC3461], [@!RFC6533]) and JSON string encoding applied.
```

Comment:
RFC 5321, [section 4.1.2](https://tools.ietf.org/html/rfc5321#section-4.1.2) defines a SMTP parameter as

    esmtp-param    = esmtp-keyword ["=" esmtp-value]

I guess a SMTP parameter without value is encoded in JMAP as an object key with value null?

Reply:
Yes. I'll add this explicitly to the spec.